### PR TITLE
Added Holiday Provider for Brazil

### PR DESCRIPTION
**Pull request by dorianneto**

### CODE REVIEW

**Comment by stelgenhof** [comment]
Thank you @dorianneto ! Very nice, however it has some conflicts with the current branch. Can you check again?

**Comment by dorianneto** [comment]
Checked and validated :)

I'm sorry for a lot of commits. I forgot to use the `php-cs-fixer` before make push :P

**Comment by dorianneto** [comment]
@stelgenhof ?

**Comment by stelgenhof** [comment]
@dorianneto Not sure what you mean, though :) I still need to check your PR; planning this for the 1.4 version. Once I have finished 1.3 I will merge it into the master branch. In case you need it earlier, you can still use your own fork.

File: tests/Brazil/BrazilBaseTestCase.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * Country (name) to be tested
+     */
+    const REGION = 'Brazil';
+
+    /**
+     * Timezone in which this provider has holidays defined
+     */
+    const TIMEZONE = 'America/Fortaleza';
+}
```
Comment (stelgenhof) [inline tests/Brazil/BrazilBaseTestCase.php]: Please add a constant for the common locale in Brazil. See for example the other test classes (    `const LOCALE = 'nl_NL';`)

File: tests/Brazil/CarnavalTest.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+namespace Yasumi\tests\Brazil;
+
+use DateTime;
+use DateInterval;
+use Yasumi\Provider\ChristianHolidays;
+
+/**
+ * Class for testing Carnaval in the Brazil.
+ */
+class CarnavalTest extends BrazilBaseTestCase
```
Comment (stelgenhof) [inline tests/Brazil/CarnavalTest.php]: All holiday test classes need to implement the contract `YasumiTestCaseInterface`. This to ensure some necessary test methods are included. Please add these for all the holiday test classes.

File: src/Yasumi/Provider/Brazil.php [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+         *
+         * Carnaval do Brasil é a maior festa popular do país. A festa acontece durante quatro dias. (que precedem a
+         * quarta–feira de cinzas). O último dia de Carnaval precede a quarta-feira de cinzas (início da Quaresma).
+         *
+         * @link https://pt.wikipedia.org/wiki/Carnaval_do_Brasil
+         */
+        if ($this->year >= 1700) {
+            $this->addHoliday(new Holiday('carnaval', [],
+                $this->calculateEaster($this->year, $this->timezone)->sub(new DateInterval('P47D')), $this->locale));
+        }
```
Comment (stelgenhof) [inline src/Yasumi/Provider/Brazil.php]: I believe Carnaval is an observed holiday, not an official/national (by law) holiday. If so, please add the appropriate parameter `HOLIDAY::TYPE_OBSERVANCE` to the **addHoliday** method.

File: src/Yasumi/Provider/Brazil.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    {
+        $this->timezone = 'America/Fortaleza';
+
+        // Add common holidays
+        $this->addHoliday($this->newYearsDay($this->year, $this->timezone, $this->locale));
+        $this->addHoliday($this->internationalWorkersDay($this->year, $this->timezone, $this->locale));
+
+        // Add Christian holidays
+        $this->addHoliday($this->christmasDay($this->year, $this->timezone, $this->locale));
+        $this->addHoliday($this->easter($this->year, $this->timezone, $this->locale));
```
Comment (stelgenhof) [inline src/Yasumi/Provider/Brazil.php]: I believe Easter is an observed holiday in Brazil, not an official/national (by law) holiday. If so, please add the appropriate parameter `HOLIDAY::TYPE_OBSERVANCE` to the **addHoliday** method.

File: src/Yasumi/Provider/Brazil.php [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+        $this->addHoliday($this->corpusChristi($this->year, $this->timezone, $this->locale));
+        $this->addHoliday($this->goodFriday($this->year, $this->timezone, $this->locale));
+
+        /**
+         * Carnaval
```
Comment (stelgenhof) [inline src/Yasumi/Provider/Brazil.php]: Would you mind providing English description for these holidays? :) Most of the users are English speaking and would not necessarily understand Brazilian Portuguese.

File: src/Yasumi/Provider/Brazil.php [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        $this->timezone = 'America/Fortaleza';
+
+        // Add common holidays
+        $this->addHoliday($this->newYearsDay($this->year, $this->timezone, $this->locale));
+        $this->addHoliday($this->internationalWorkersDay($this->year, $this->timezone, $this->locale));
+
+        // Add Christian holidays
+        $this->addHoliday($this->christmasDay($this->year, $this->timezone, $this->locale));
+        $this->addHoliday($this->easter($this->year, $this->timezone, $this->locale));
+        $this->addHoliday($this->corpusChristi($this->year, $this->timezone, $this->locale));
```
Comment (stelgenhof) [inline src/Yasumi/Provider/Brazil.php]: I believe Corpus Christi is an optional holiday in Brazil, not an official/national (by law) holiday. If so, please add the appropriate parameter `HOLIDAY::TYPE_OBSERVANCE` to the **addHoliday** method.

File: tests/Brazil/BrazilBaseTestCase.php (hunk 1 above)
Comment (dorianneto) [inline tests/Brazil/BrazilBaseTestCase.php]: done!

File: tests/Brazil/CarnavalTest.php (hunk 2 above)
Comment (dorianneto) [inline tests/Brazil/CarnavalTest.php]: done!

File: src/Yasumi/Provider/Brazil.php (hunk 5 above)
Comment (dorianneto) [inline src/Yasumi/Provider/Brazil.php]: done!

File: src/Yasumi/Provider/Brazil.php (hunk 3 above)
Comment (dorianneto) [inline src/Yasumi/Provider/Brazil.php]: done!

File: src/Yasumi/Provider/Brazil.php (hunk 4 above)
Comment (dorianneto) [inline src/Yasumi/Provider/Brazil.php]: done!

File: src/Yasumi/Provider/Brazil.php (hunk 6 above)
Comment (dorianneto) [inline src/Yasumi/Provider/Brazil.php]: done!

**Comment by dorianneto** [comment]
All done :)

**Comment by stelgenhof** [comment]
Thanks @dorianneto ! Looking good. Let me review once more and merge it soon.
Appreciate the PR!

**Comment by stelgenhof** [comment]
Just ran another unittest and found some errors:

```
1) Yasumi\tests\Brazil\BrazilTest::testNationalHolidays
Failed asserting that an array has the key 'tiradentesDay'.
```

If a holidays is of a certain type (e.g. National), make sure they are added in the respective test of the BrazilTest class. 

Running a single iteration of PHPUnit might not detect all issues. Please run with --repeat 10 for example. You might see more issues appearing.

File: tests/Brazil/BrazilTest.php [hunk 7]
```diff
@@ -31,16 +31,13 @@ public function testNationalHolidays()
     {
         $this->assertDefinedHolidays([
             'newYearsDay',
-            'carnaval',
             'goodFriday',
-            'easter',
-            'diaDeTiradentes',
+            'tiradentesDay',
```
Comment (dorianneto) [inline tests/Brazil/BrazilTest.php]: The holiday is in the correct array.

**Comment by dorianneto** [comment]
Before push this commit, I verified all unittests and not found no problem. Now, I ran with --repeat 10 and sometimes have a problem sometimes not.

This is a PHPUnit bug or I have to do anything? I never experienced  this problem, sorry :(

**Comment by stelgenhof** [comment]
The reason for that is, that the unit tests use a randomized year. It might be possible that for that particular year the respective holiday doesn't exist. Usually this becomes apparent when running multiple iterations :). Nothing to worry about. Let me merge it and fix it for you.

**Pull request merged**